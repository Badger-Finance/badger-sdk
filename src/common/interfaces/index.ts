export { BlocksRangeOptions } from './blocks-range-options.interface';
export { RangeOptions } from './range-options.interface';
export { TimeRangeOptions } from './time-range-options.interface';
