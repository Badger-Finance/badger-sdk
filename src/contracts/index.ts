/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { BadgerTree } from "./BadgerTree";
export type { Digg } from "./Digg";
export type { Erc20 } from "./Erc20";
export type { Ibbtc } from "./Ibbtc";
export type { Registry } from "./Registry";
export type { RewardsLogger } from "./RewardsLogger";
export type { Sett } from "./Sett";
export type { StrategyV15 } from "./StrategyV15";
export type { VaultV15 } from "./VaultV15";

export { BadgerTree__factory } from "./factories/BadgerTree__factory";
export { Digg__factory } from "./factories/Digg__factory";
export { Erc20__factory } from "./factories/Erc20__factory";
export { Ibbtc__factory } from "./factories/Ibbtc__factory";
export { Registry__factory } from "./factories/Registry__factory";
export { RewardsLogger__factory } from "./factories/RewardsLogger__factory";
export { Sett__factory } from "./factories/Sett__factory";
export { StrategyV15__factory } from "./factories/StrategyV15__factory";
export { VaultV15__factory } from "./factories/VaultV15__factory";
