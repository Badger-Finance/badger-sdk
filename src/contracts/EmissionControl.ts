/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import { FunctionFragment, Result, EventFragment } from "@ethersproject/abi";
import { Listener, Provider } from "@ethersproject/providers";
import { TypedEventFilter, TypedEvent, TypedListener, OnEvent } from "./common";

export interface EmissionControlInterface extends utils.Interface {
  contractName: "EmissionControl";
  functions: {
    "MAX_BPS()": FunctionFragment;
    "addManager(address)": FunctionFragment;
    "boostedEmissionRate(address)": FunctionFragment;
    "manager(address)": FunctionFragment;
    "owner()": FunctionFragment;
    "proRataEmissionRate(address)": FunctionFragment;
    "removeManager(address)": FunctionFragment;
    "renounceOwnership()": FunctionFragment;
    "setBoostedEmission(address,uint256)": FunctionFragment;
    "setTokenWeight(address,uint256)": FunctionFragment;
    "tokenWeight(address)": FunctionFragment;
    "transferOwnership(address)": FunctionFragment;
  };

  encodeFunctionData(functionFragment: "MAX_BPS", values?: undefined): string;
  encodeFunctionData(functionFragment: "addManager", values: [string]): string;
  encodeFunctionData(
    functionFragment: "boostedEmissionRate",
    values: [string]
  ): string;
  encodeFunctionData(functionFragment: "manager", values: [string]): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "proRataEmissionRate",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "removeManager",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "renounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setBoostedEmission",
    values: [string, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setTokenWeight",
    values: [string, BigNumberish]
  ): string;
  encodeFunctionData(functionFragment: "tokenWeight", values: [string]): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [string]
  ): string;

  decodeFunctionResult(functionFragment: "MAX_BPS", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "addManager", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "boostedEmissionRate",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "manager", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "proRataEmissionRate",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "removeManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setBoostedEmission",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setTokenWeight",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "tokenWeight",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;

  events: {
    "OwnershipTransferred(address,address)": EventFragment;
    "TokenBoostedEmissionChanged(address,uint256)": EventFragment;
    "TokenWeightChanged(address,uint256)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "OwnershipTransferred"): EventFragment;
  getEvent(
    nameOrSignatureOrTopic: "TokenBoostedEmissionChanged"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "TokenWeightChanged"): EventFragment;
}

export type OwnershipTransferredEvent = TypedEvent<
  [string, string],
  { previousOwner: string; newOwner: string }
>;

export type OwnershipTransferredEventFilter =
  TypedEventFilter<OwnershipTransferredEvent>;

export type TokenBoostedEmissionChangedEvent = TypedEvent<
  [string, BigNumber],
  { _vault: string; _weight: BigNumber }
>;

export type TokenBoostedEmissionChangedEventFilter =
  TypedEventFilter<TokenBoostedEmissionChangedEvent>;

export type TokenWeightChangedEvent = TypedEvent<
  [string, BigNumber],
  { _token: string; _weight: BigNumber }
>;

export type TokenWeightChangedEventFilter =
  TypedEventFilter<TokenWeightChangedEvent>;

export interface EmissionControl extends BaseContract {
  contractName: "EmissionControl";
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: EmissionControlInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    MAX_BPS(overrides?: CallOverrides): Promise<[BigNumber]>;

    addManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    boostedEmissionRate(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    manager(arg0: string, overrides?: CallOverrides): Promise<[boolean]>;

    owner(overrides?: CallOverrides): Promise<[string]>;

    proRataEmissionRate(
      _vault: string,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    removeManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    setBoostedEmission(
      _vault: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    setTokenWeight(
      _token: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;

    tokenWeight(arg0: string, overrides?: CallOverrides): Promise<[BigNumber]>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<ContractTransaction>;
  };

  MAX_BPS(overrides?: CallOverrides): Promise<BigNumber>;

  addManager(
    _manager: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  boostedEmissionRate(
    arg0: string,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  manager(arg0: string, overrides?: CallOverrides): Promise<boolean>;

  owner(overrides?: CallOverrides): Promise<string>;

  proRataEmissionRate(
    _vault: string,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  removeManager(
    _manager: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  renounceOwnership(
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  setBoostedEmission(
    _vault: string,
    _weight: BigNumberish,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  setTokenWeight(
    _token: string,
    _weight: BigNumberish,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  tokenWeight(arg0: string, overrides?: CallOverrides): Promise<BigNumber>;

  transferOwnership(
    newOwner: string,
    overrides?: Overrides & { from?: string | Promise<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    MAX_BPS(overrides?: CallOverrides): Promise<BigNumber>;

    addManager(_manager: string, overrides?: CallOverrides): Promise<void>;

    boostedEmissionRate(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    manager(arg0: string, overrides?: CallOverrides): Promise<boolean>;

    owner(overrides?: CallOverrides): Promise<string>;

    proRataEmissionRate(
      _vault: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    removeManager(_manager: string, overrides?: CallOverrides): Promise<void>;

    renounceOwnership(overrides?: CallOverrides): Promise<void>;

    setBoostedEmission(
      _vault: string,
      _weight: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    setTokenWeight(
      _token: string,
      _weight: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    tokenWeight(arg0: string, overrides?: CallOverrides): Promise<BigNumber>;

    transferOwnership(
      newOwner: string,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "OwnershipTransferred(address,address)"(
      previousOwner?: string | null,
      newOwner?: string | null
    ): OwnershipTransferredEventFilter;
    OwnershipTransferred(
      previousOwner?: string | null,
      newOwner?: string | null
    ): OwnershipTransferredEventFilter;

    "TokenBoostedEmissionChanged(address,uint256)"(
      _vault?: string | null,
      _weight?: BigNumberish | null
    ): TokenBoostedEmissionChangedEventFilter;
    TokenBoostedEmissionChanged(
      _vault?: string | null,
      _weight?: BigNumberish | null
    ): TokenBoostedEmissionChangedEventFilter;

    "TokenWeightChanged(address,uint256)"(
      _token?: string | null,
      _weight?: BigNumberish | null
    ): TokenWeightChangedEventFilter;
    TokenWeightChanged(
      _token?: string | null,
      _weight?: BigNumberish | null
    ): TokenWeightChangedEventFilter;
  };

  estimateGas: {
    MAX_BPS(overrides?: CallOverrides): Promise<BigNumber>;

    addManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    boostedEmissionRate(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    manager(arg0: string, overrides?: CallOverrides): Promise<BigNumber>;

    owner(overrides?: CallOverrides): Promise<BigNumber>;

    proRataEmissionRate(
      _vault: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    removeManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    setBoostedEmission(
      _vault: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    setTokenWeight(
      _token: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;

    tokenWeight(arg0: string, overrides?: CallOverrides): Promise<BigNumber>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    MAX_BPS(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    addManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    boostedEmissionRate(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    manager(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    owner(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    proRataEmissionRate(
      _vault: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    removeManager(
      _manager: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    renounceOwnership(
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    setBoostedEmission(
      _vault: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    setTokenWeight(
      _token: string,
      _weight: BigNumberish,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;

    tokenWeight(
      arg0: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    transferOwnership(
      newOwner: string,
      overrides?: Overrides & { from?: string | Promise<string> }
    ): Promise<PopulatedTransaction>;
  };
}
