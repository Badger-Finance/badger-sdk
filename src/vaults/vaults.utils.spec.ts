import { BigNumber, ethers } from 'ethers';

import { VaultState } from '..';
import { TimeRangeOptions } from '../common';
import { Strategy__factory, VaultV15__factory } from '../contracts';
import { HarvestEvent, TreeDistributionEvent } from '../contracts/Strategy';
import {
  HarvestedEvent,
  TreeDistributionEvent as TreeDistributionEventV15,
} from '../contracts/VaultV15';
import { VaultVersion } from '.';
import {
  getVaultState,
  getVaultVersion,
  loadVaultPerformanceEvents,
  loadVaultV15PerformanceEvents,
  parseHarvestEvents,
  parseHarvestV15Events,
  timestampInRange,
} from './vaults.utils';

describe('vaults.utils', () => {
  const defaultTimestamp = BigNumber.from('1647284957');
  const block = {
    hash: '',
    parentHash: '',
    number: 1_000_000,
    timestamp: defaultTimestamp.toNumber(),
    nonce: '',
    difficulty: 0,
    _difficulty: BigNumber.from('0'),
    gasLimit: ethers.constants.Zero,
    gasUsed: ethers.constants.Zero,
    miner: '',
    extraData: '',
    transactions: [],
  };

  // harvest one
  const harvestedOne = BigNumber.from('1647275665000000');
  const blockNumberOne = BigNumber.from('1647275665');

  // harvest two
  const harvestedTwo = BigNumber.from('1447275665000000');
  const blockNumberTwo = BigNumber.from('1647315665');

  // harvest three
  const harvestedThree = BigNumber.from('1947275665000000');
  const blockNumberThree = BigNumber.from('1647355665');

  // setup harvest events
  const harvests: HarvestEvent[] = [
    { args: [harvestedOne, blockNumberOne] },
    { args: [harvestedTwo, blockNumberTwo] },
    { args: [harvestedThree, blockNumberThree] },
  ] as HarvestEvent[];
  harvests.forEach((h, i) => {
    h.getBlock = async () => ({
      ...block,
      timestamp: block.timestamp + 1250 * i,
    });
    h.transactionHash = `0x6422091f26311d0442aa546f3f1444bcc111e361cbd506659f38082b1edab172${i}`;
  });

  const harvestsV15: HarvestedEvent[] = [
    {
      args: ['0xTEST', harvestedOne, blockNumberOne, defaultTimestamp],
      transactionHash:
        '0x6422091f26311d0442aa546f3f1444bcc111e361cbd506659f38082b1edab172',
    },
    {
      args: [
        '0xTEST',
        harvestedTwo,
        blockNumberTwo,
        defaultTimestamp.add(1250),
      ],
      transactionHash:
        '0x6422091f26311d0442aa546f3f1444bcc111e361cbd506659f38082b1edab172',
    },
    {
      args: [
        '0xTEST',
        harvestedThree,
        blockNumberThree,
        defaultTimestamp.add(2500),
      ],
      transactionHash:
        '0x30bc2ab3a59f7923ea20f7b99331dbc974130dc8b7152bb897d393fc2c506214',
    },
  ] as HarvestedEvent[];

  // distibution one
  const distributedOne = BigNumber.from('1675665000000');

  // distibution three (skip two)
  const distributedThree = BigNumber.from('1947275665000000');

  const distributions: TreeDistributionEvent[] = [
    {
      args: ['0xBTC', distributedOne, blockNumberOne, defaultTimestamp],
      transactionHash:
        '0x79ac9d3c0acfb293b07d539df3714cd873fc839e572894e346fea91c59b20afb',
    },
    {
      args: [
        '0xBTC',
        distributedThree,
        blockNumberThree,
        defaultTimestamp.add(1250),
      ],
      transactionHash:
        '0x79ac9d3c0acfb293b07d539df3714cd873fc839e572894e346fea91c59b20afb',
    },
  ] as TreeDistributionEvent[];

  const distributionsV15: TreeDistributionEventV15[] = [
    {
      args: ['0xBTC', distributedOne, blockNumberOne, defaultTimestamp],
      transactionHash:
        '0x93286e1f1405cdc8441fe0102dd8663e0d13eea88e032743e287c365cfe290e5',
    },
    {
      args: [
        '0xBTC',
        distributedThree,
        blockNumberThree,
        defaultTimestamp.add(1750),
      ],
      transactionHash:
        '0x93286e1f1405cdc8441fe0102dd8663e0d13eea88e032743e287c365cfe290e5',
    },
  ] as TreeDistributionEventV15[];

  describe('parseHarvestEvents', () => {
    it('converts vault events to typed usable events', async () => {
      const result = await parseHarvestEvents(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        harvests,
        distributions,
      );
      expect(result).toMatchSnapshot();
    });

    it('returns zero for timestamp on getBlock error', async () => {
      harvests.forEach(
        (h, i) =>
          (h.getBlock = async () => {
            if (i % 2 === 1) {
              throw new Error('Expected test error: getBlock');
            }
            return {
              ...block,
              timestamp: block.timestamp + 1250 * i,
            };
          }),
      );
      const result = await parseHarvestEvents(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        harvests,
        distributions,
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('parseHarvestV15Events', () => {
    it('converts vault events to typed usable events', async () => {
      const result = await parseHarvestV15Events(harvestsV15, distributionsV15);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getVaultVersion', () => {
    it.each([
      ['v1', VaultVersion.v1],
      ['v1.5', VaultVersion.v1_5],
      ['v2', VaultVersion.v2],
      ['v.bad.rc2', VaultVersion.v1],
    ])(
      'evaluates %s as vault version %s',
      (input: string, version: VaultVersion) => {
        expect(getVaultVersion(input)).toEqual(version);
      },
    );
  });

  describe('getVaultState', () => {
    it.each([
      [0, VaultState.Experimental],
      [1, VaultState.Guarded],
      [2, VaultState.Open],
      [-1, VaultState.Experimental],
      [3, VaultState.Experimental],
    ])(
      'evaluates %s as vault state %s',
      (input: number, version: VaultState) => {
        expect(getVaultState(input)).toEqual(version);
      },
    );
  });

  describe('timestampInRange', () => {
    const defaultTimestamp = Date.now();

    it.each([
      [{}, defaultTimestamp, true],
      [{ timestamp_gte: defaultTimestamp }, defaultTimestamp, true],
      [{ timestamp_gte: defaultTimestamp + 1 }, defaultTimestamp, false],
      [{ timestamp_gt: defaultTimestamp - 1 }, defaultTimestamp, true],
      [{ timestamp_gt: defaultTimestamp }, defaultTimestamp, false],
      [{ timestamp_lte: defaultTimestamp }, defaultTimestamp, true],
      [{ timestamp_lte: defaultTimestamp - 1 }, defaultTimestamp, false],
      [{ timestamp_lt: defaultTimestamp + 1 }, defaultTimestamp, true],
      [{ timestamp_lt: defaultTimestamp }, defaultTimestamp, false],
      [
        { timestamp_lte: defaultTimestamp, timestamp_gte: defaultTimestamp },
        defaultTimestamp,
        true,
      ],
      [
        {
          timestamp_lt: defaultTimestamp + 1,
          timestamp_gt: defaultTimestamp - 1,
        },
        defaultTimestamp,
        true,
      ],
      [
        { timestamp_lt: defaultTimestamp + 3, timestamp_gt: defaultTimestamp },
        defaultTimestamp,
        false,
      ],
      [
        { timestamp_lt: defaultTimestamp, timestamp_gt: defaultTimestamp - 3 },
        defaultTimestamp,
        false,
      ],
    ])(
      'evaluates %p options with %d timestamp as %s',
      (options: TimeRangeOptions, timestamp: number, expected: boolean) => {
        expect(timestampInRange(options, timestamp)).toEqual(expected);
      },
    );

    it('throws an error on an invalid time range', () => {
      const upper = defaultTimestamp + 1;
      expect(() => {
        timestampInRange(
          {
            timestamp_gte: upper,
            timestamp_lte: defaultTimestamp,
          },
          defaultTimestamp,
        );
      }).toThrow(
        `Invalid time range check requested (${upper} - ${defaultTimestamp})`,
      );
    });
  });

  describe('loadVaultPerformanceEvents', () => {
    it('loads recent vault harvests and tree distributions', async () => {
      const testStrategy = Strategy__factory.connect(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        new ethers.providers.JsonRpcProvider(''),
      );
      jest
        .spyOn(testStrategy, 'queryFilter')
        .mockImplementation(async (filter) => {
          if (!filter.topics || typeof filter.topics === 'string') {
            return [];
          }
          const topic = filter.topics[0];
          if (
            topic ===
            '0x6c8433a8e155f0af04dba058d4e4695f7da554578963d876bdf4a6d8d6399d9c'
          ) {
            return harvests;
          }
          return distributions;
        });
      const result = await loadVaultPerformanceEvents(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        testStrategy,
        {
          startBlock: 0,
          endBlock: 100,
        },
      );
      expect(result).toMatchSnapshot();
    });

    it('loads no events if not available', async () => {
      const testStrategy = Strategy__factory.connect(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        new ethers.providers.JsonRpcProvider(''),
      );
      jest
        .spyOn(testStrategy, 'queryFilter')
        .mockImplementation(async () => []);
      const result = await loadVaultPerformanceEvents(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        testStrategy,
        {
          startBlock: 0,
          endBlock: 100,
        },
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('loadVaultV15PerformanceEvents', () => {
    it('loads recent vault harvests and tree distributions', async () => {
      const testVault = VaultV15__factory.connect(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        new ethers.providers.JsonRpcProvider(''),
      );
      jest
        .spyOn(testVault, 'queryFilter')
        .mockImplementation(async (filter) => {
          if (!filter.topics || typeof filter.topics === 'string') {
            return [];
          }
          const topic = filter.topics[0];
          if (
            topic ===
            '0xe48bba143e2a0b557fa6f3234bd6ffc704518cc98c7de6a2385549fae27d1b75'
          ) {
            return harvestsV15;
          }
          return distributionsV15;
        });
      const result = await loadVaultV15PerformanceEvents(testVault, {
        startBlock: 0,
        endBlock: 100,
      });
      expect(result).toMatchSnapshot();
    });

    it('loads no events if not available', async () => {
      const testVault = VaultV15__factory.connect(
        '0x1ccca1ce62c62f7be95d4a67722a8fdbed6eecb4',
        new ethers.providers.JsonRpcProvider(''),
      );
      jest.spyOn(testVault, 'queryFilter').mockImplementation(async () => []);
      const result = await loadVaultV15PerformanceEvents(testVault, {
        startBlock: 0,
        endBlock: 100,
      });
      expect(result).toMatchSnapshot();
    });
  });
});
