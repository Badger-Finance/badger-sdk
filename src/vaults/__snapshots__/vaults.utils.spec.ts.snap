// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vaults.utils loadVaultPerformanceEvents loads no events if not available 1`] = `
Object {
  "data": Array [],
}
`;

exports[`vaults.utils loadVaultPerformanceEvents loads recent vault harvests and tree distributions 1`] = `
Object {
  "data": Array [
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x05244a1fe78640",
            "type": "BigNumber",
          },
          "block": undefined,
          "timestamp": 0,
          "token": "",
        },
      ],
      "timestamp": 0,
      "treeDistributions": Array [],
    },
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x05da3040dc0640",
            "type": "BigNumber",
          },
          "block": 1000000,
          "timestamp": 1647284957,
          "token": "",
        },
      ],
      "timestamp": 1647284957,
      "treeDistributions": Array [
        Object {
          "amount": Object {
            "hex": "0x0186256ac640",
            "type": "BigNumber",
          },
          "block": 1647275665,
          "timestamp": 1647284957,
          "token": "0xBTC",
        },
      ],
    },
    Object {
      "harvests": Array [],
      "timestamp": 1647286207,
      "treeDistributions": Array [
        Object {
          "amount": Object {
            "hex": "0x06eb09724ac640",
            "type": "BigNumber",
          },
          "block": 1647355665,
          "timestamp": 1647286207,
          "token": "0xBTC",
        },
      ],
    },
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x06eb09724ac640",
            "type": "BigNumber",
          },
          "block": 1000000,
          "timestamp": 1647287457,
          "token": "",
        },
      ],
      "timestamp": 1647287457,
      "treeDistributions": Array [],
    },
  ],
}
`;

exports[`vaults.utils loadVaultV15PerformanceEvents loads no events if not available 1`] = `
Object {
  "data": Array [],
}
`;

exports[`vaults.utils loadVaultV15PerformanceEvents loads recent vault harvests and tree distributions 1`] = `
Object {
  "data": Array [
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x05da3040dc0640",
            "type": "BigNumber",
          },
          "block": 1647275665,
          "timestamp": 1647284957,
          "token": "0xTEST",
        },
      ],
      "timestamp": 1647284957,
      "treeDistributions": Array [
        Object {
          "amount": Object {
            "hex": "0x0186256ac640",
            "type": "BigNumber",
          },
          "block": 1647275665,
          "timestamp": 1647284957,
          "token": "0xBTC",
        },
      ],
    },
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x05244a1fe78640",
            "type": "BigNumber",
          },
          "block": 1647315665,
          "timestamp": 1647286207,
          "token": "0xTEST",
        },
      ],
      "timestamp": 1647286207,
      "treeDistributions": Array [],
    },
    Object {
      "harvests": Array [],
      "timestamp": 1647286707,
      "treeDistributions": Array [
        Object {
          "amount": Object {
            "hex": "0x06eb09724ac640",
            "type": "BigNumber",
          },
          "block": 1647355665,
          "timestamp": 1647286707,
          "token": "0xBTC",
        },
      ],
    },
    Object {
      "harvests": Array [
        Object {
          "amount": Object {
            "hex": "0x06eb09724ac640",
            "type": "BigNumber",
          },
          "block": 1647355665,
          "timestamp": 1647287457,
          "token": "0xTEST",
        },
      ],
      "timestamp": 1647287457,
      "treeDistributions": Array [],
    },
  ],
}
`;

exports[`vaults.utils parseHarvestEvents converts vault events to typed usable events 1`] = `
Object {
  "distributions": Array [
    Object {
      "amount": Object {
        "hex": "0x0186256ac640",
        "type": "BigNumber",
      },
      "block": 1647275665,
      "timestamp": 1647284957,
      "token": "0xBTC",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1647355665,
      "timestamp": 1647286207,
      "token": "0xBTC",
    },
  ],
  "harvests": Array [
    Object {
      "amount": Object {
        "hex": "0x05da3040dc0640",
        "type": "BigNumber",
      },
      "block": 1000000,
      "timestamp": 1647284957,
      "token": "",
    },
    Object {
      "amount": Object {
        "hex": "0x05244a1fe78640",
        "type": "BigNumber",
      },
      "block": 1000000,
      "timestamp": 1647286207,
      "token": "",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1000000,
      "timestamp": 1647287457,
      "token": "",
    },
  ],
}
`;

exports[`vaults.utils parseHarvestEvents returns zero for timestamp on getBlock error 1`] = `
Object {
  "distributions": Array [
    Object {
      "amount": Object {
        "hex": "0x0186256ac640",
        "type": "BigNumber",
      },
      "block": 1647275665,
      "timestamp": 1647284957,
      "token": "0xBTC",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1647355665,
      "timestamp": 1647286207,
      "token": "0xBTC",
    },
  ],
  "harvests": Array [
    Object {
      "amount": Object {
        "hex": "0x05da3040dc0640",
        "type": "BigNumber",
      },
      "block": 1000000,
      "timestamp": 1647284957,
      "token": "",
    },
    Object {
      "amount": Object {
        "hex": "0x05244a1fe78640",
        "type": "BigNumber",
      },
      "block": undefined,
      "timestamp": 0,
      "token": "",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1000000,
      "timestamp": 1647287457,
      "token": "",
    },
  ],
}
`;

exports[`vaults.utils parseHarvestV15Events converts vault events to typed usable events 1`] = `
Object {
  "distributions": Array [
    Object {
      "amount": Object {
        "hex": "0x0186256ac640",
        "type": "BigNumber",
      },
      "block": 1647275665,
      "timestamp": 1647284957,
      "token": "0xBTC",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1647355665,
      "timestamp": 1647286707,
      "token": "0xBTC",
    },
  ],
  "harvests": Array [
    Object {
      "amount": Object {
        "hex": "0x05da3040dc0640",
        "type": "BigNumber",
      },
      "block": 1647275665,
      "timestamp": 1647284957,
      "token": "0xTEST",
    },
    Object {
      "amount": Object {
        "hex": "0x05244a1fe78640",
        "type": "BigNumber",
      },
      "block": 1647315665,
      "timestamp": 1647286207,
      "token": "0xTEST",
    },
    Object {
      "amount": Object {
        "hex": "0x06eb09724ac640",
        "type": "BigNumber",
      },
      "block": 1647355665,
      "timestamp": 1647287457,
      "token": "0xTEST",
    },
  ],
}
`;
